sdl2promela/stop_condition/translator.py

```diff
@@ -66,173 +90,202 @@ def _generate(expr: model.XorExpression):
+    result = (
         promelaBuilder.BinaryExpressionBuilder(promela.BinaryOperator.MULTIPLY)
-        .withLeft(_generate(expr.lhs))
-        .withRight(_generate(expr.rhs))
+        .withLeft(_generate(context, expr.lhs))
+        .withRight(_generate(context, expr.rhs))
         .build()
     )
+    context.clear()
+    return result
 
 
-@dispatch(model.DivExpression)
-def _generate(expr: model.DivExpression):
-    return (
+@dispatch(GenerateContext, model.DivExpression)
+def _generate(context: GenerateContext, expr: model.DivExpression):
+    result = (
         promelaBuilder.BinaryExpressionBuilder(promela.BinaryOperator.DIVIDE)
-        .withLeft(_generate(expr.lhs))
-        .withRight(_generate(expr.rhs))
+        .withLeft(_generate(context, expr.lhs))
+        .withRight(_generate(context, expr.rhs))
         .build()
     )
+    context.clear()
+    return result
 
 
-@dispatch(model.ModExpression)
-def _generate(expr: model.ModExpression):
-    return (
+@dispatch(GenerateContext, model.ModExpression)
+def _generate(context: GenerateContext, expr: model.ModExpression):
+    result = (
         promelaBuilder.BinaryExpressionBuilder(promela.BinaryOperator.MODULO)
-        .withLeft(_generate(expr.lhs))
-        .withRight(_generate(expr.rhs))
+        .withLeft(_generate(context, expr.lhs))
+        .withRight(_generate(context, expr.rhs))
         .build()
     )
+    context.clear()
+    return result
 
 
-@dispatch(model.NegationExpression)
-def _generate(expr: model.NegationExpression):
-    return (
+@dispatch(GenerateContext, model.NegationExpression)
+def _generate(context: GenerateContext, expr: model.NegationExpression):
+    result = (
         promelaBuilder.UnaryExpressionBuilder(promela.UnaryOperator.NEGATIVE)
-        .withExpression(expr.expression)
+        .withExpression(_generate(context, expr.expression))
         .build()
     )
+    context.clear()
+    return result
 
 
-@dispatch(model.IntegerValue)
-def _generate(expr: model.IntegerValue):
+@dispatch(GenerateContext, model.IntegerValue)
+def _generate(context: GenerateContext, expr: model.IntegerValue):
     return promela.IntegerValue(expr.value)
 
 
-@dispatch(model.FloatValue)
-def _generate(expr: model.FloatValue):
+@dispatch(GenerateContext, model.FloatValue)
+def _generate(context: GenerateContext, expr: model.FloatValue):
     return promela.FloatValue(expr.value)
 
 
-@dispatch(model.BooleanValue)
-def _generate(expr: model.BooleanValue):
+@dispatch(GenerateContext, model.BooleanValue)
+def _generate(context: GenerateContext, expr: model.BooleanValue):
     return promela.BooleanValue(expr.value)
 
 
-@dispatch(model.VariableReference)
-def _generate(expr: model.VariableReference):
-    return promelaBuilder.VariableReferenceBuilder(expr.name).build()
-
+@dispatch(GenerateContext, model.VariableReference)
+def _generate(context: GenerateContext, expr: model.VariableReference):
+    if context.choice_selection is not None:
```

Comment:
This really asks for a description, or maybe more descriptive name?

Reply:
code commented. I cannot rename this function, it uses 'dispatch' the name shall be consistant with the others.